File: snobot_sim_java/src/main/java/com/snobot/simulator/simulator_components/ctre/CtrePigeonImuSim.java
package com.snobot.simulator.simulator_components.ctre;

import com.snobot.simulator.SensorActuatorRegistry;
import com.snobot.simulator.SimDeviceHelper;
import com.snobot.simulator.module_wrapper.ASensorWrapper;
import com.snobot.simulator.module_wrapper.interfaces.IAccelerometerWrapper;
import com.snobot.simulator.module_wrapper.interfaces.IGyroWrapper;
import edu.wpi.first.hal.SimDouble;
import edu.wpi.first.hal.sim.SimDeviceSim;

public class CtrePigeonImuSim extends ASensorWrapper
{
    public static final int sCTRE_OFFSET = 400;

    private final PigeonAccelWrapper mXAccel;
    private final PigeonAccelWrapper mYAccel;
    private final PigeonAccelWrapper mZAccel;

    private final PigeonGyroWrapper mYawGyro;
    private final PigeonGyroWrapper mPitchGyro;
    private final PigeonGyroWrapper mRollGyro;

    private final int mCanHandle;

    ///////////////////////////////////////////////
    private SimDouble mYawAngleDeg;
    private SimDouble mRawGyroXyzDps0;
    private SimDouble mRawGyroXyzDps1;
    private SimDouble mRawGyroXyzDps2;
    private SimDouble mYawPitchRollYpr0;
    private SimDouble mYawPitchRollYpr1;
    private SimDouble mYawPitchRollYpr2;
    private SimDouble mFusedHeading2Value;
    ///////////////////////////////////////////////

    public CtrePigeonImuSim(int aCanHandle, int aBasePort)
    {
        super("Pigeon IMU");

        mCanHandle = aCanHandle;



        mXAccel = new PigeonAccelWrapper("Pigeon X");
        mYAccel = new PigeonAccelWrapper("Pigeon Y");
        mZAccel = new PigeonAccelWrapper("Pigeon Z");

        mYawGyro = new PigeonGyroWrapper("Pigeon Yaw");
        mPitchGyro = new PigeonGyroWrapper("Pigeon Pitch");
        mRollGyro = new PigeonGyroWrapper("Pigeon Roll");

        SensorActuatorRegistry.get().register(mXAccel, aBasePort + 0);
        SensorActuatorRegistry.get().register(mYAccel, aBasePort + 1);
        SensorActuatorRegistry.get().register(mZAccel, aBasePort + 2);

        SensorActuatorRegistry.get().register(mYawGyro, aBasePort + 0);
        SensorActuatorRegistry.get().register(mPitchGyro, aBasePort + 1);
        SensorActuatorRegistry.get().register(mRollGyro, aBasePort + 2);
    }


    @Override
    public void setInitialized(boolean aInitialized)
    {
        super.setInitialized(aInitialized);

        mXAccel.setInitialized(aInitialized);
        mYAccel.setInitialized(aInitialized);
        mZAccel.setInitialized(aInitialized);

        mYawGyro.setInitialized(aInitialized);
        mPitchGyro.setInitialized(aInitialized);
        mRollGyro.setInitialized(aInitialized);

        int deviceId = mCanHandle;
        String deviceName = "CtrePigeonIMUWrapper " + deviceId + "[" + deviceId + "]";
        System.out.println("----------------- Initializing " + deviceName);
        SimDeviceSim simDevice = new SimDeviceSim(deviceName);

        mYawAngleDeg = SimDeviceHelper.getSimDouble(simDevice, "Yaw_angleDeg");
        mRawGyroXyzDps0 = SimDeviceHelper.getSimDouble(simDevice, "RawGyro_xyz_dps_0");
        mRawGyroXyzDps1 = SimDeviceHelper.getSimDouble(simDevice, "RawGyro_xyz_dps_1");
        mRawGyroXyzDps2 = SimDeviceHelper.getSimDouble(simDevice, "RawGyro_xyz_dps_2");
        mYawPitchRollYpr0 = SimDeviceHelper.getSimDouble(simDevice, "YawPitchRoll_ypr_0");
        mYawPitchRollYpr1 = SimDeviceHelper.getSimDouble(simDevice, "YawPitchRoll_ypr_1");
        mYawPitchRollYpr2 = SimDeviceHelper.getSimDouble(simDevice, "YawPitchRoll_ypr_2");
        mFusedHeading2Value = SimDeviceHelper.getSimDouble(simDevice, "FusedHeading1_value");
    }

    public IGyroWrapper getYawWrapper()
    {
        return mYawGyro;
    }

    public IGyroWrapper getPitchWrapper()
    {
        return mPitchGyro;
    }

    public IGyroWrapper getRollWrapper()
    {
        return mRollGyro;
    }

    public IAccelerometerWrapper getXWrapper()
    {
        return mXAccel;
    }

    public IAccelerometerWrapper getYWrapper()
    {
        return mYAccel;
    }

    public IAccelerometerWrapper getZWrapper()
    {
        return mZAccel;
    }

    public static class PigeonGyroWrapper extends ASensorWrapper implements IGyroWrapper
    {
        private double mAngle;
        private double mAngleOffset;

        public PigeonGyroWrapper(String aName)
        {
            super(aName);
        }

        public void setDesiredYaw(double aOffset)
        {
            mAngleOffset = aOffset - mAngle;
        }

        @Override
        public double getAngle()
        {
            return mAngle + mAngleOffset;
        }

        @Override
        public void setAngle(double aAngle)
        {
            mAngle = aAngle;
        }
    }

    private static class PigeonAccelWrapper extends ASensorWrapper implements IAccelerometerWrapper
    {
        private double mAccel;

        public PigeonAccelWrapper(String aName)
        {
            super(aName);
        }

        @Override
        public double getAcceleration()
        {
            return mAccel;
        }

        @Override
        public void setAcceleration(double aAccel)
        {
            mAccel = aAccel;
        }
    }

    ///////////////////////////////////////////////
    public void handleSetYaw()
    {
        ((CtrePigeonImuSim.PigeonGyroWrapper) getYawWrapper()).setDesiredYaw(mYawAngleDeg.get());
    }


    public void handleGetRawGyro()
    {
        mRawGyroXyzDps0.set(getYawWrapper().getAngle());
        mRawGyroXyzDps1.set(getPitchWrapper().getAngle());
        mRawGyroXyzDps2.set(getRollWrapper().getAngle());
    }

    public void handleGetYawPitchRoll()
    {
        mYawPitchRollYpr0.set(getYawWrapper().getAngle());
        mYawPitchRollYpr1.set(getPitchWrapper().getAngle());
        mYawPitchRollYpr2.set(getRollWrapper().getAngle());
    }

    public void handleGetFusedHeading()
    {
        mFusedHeading2Value.set(getYawWrapper().getAngle());
    }
    ///////////////////////////////////////////////

}

File: snobot_sim/src/main/native/cpp/SnobotSim/SimulatorComponents/CtreWrappers/CtrePigeonImuSim.cpp

#include "SnobotSim/SimulatorComponents/CtreWrappers/CtrePigeonImuSim.h"

#include "SnobotSim/SensorActuatorRegistry.h"
#include <iostream>

namespace
{
hal::SimDouble getSimDouble(HAL_SimDeviceHandle aDeviceHandle, std::string aName)
{
    hal::SimDouble output = HALSIM_GetSimValueHandle(aDeviceHandle, aName.c_str());
    if (!output)
    {
//        throw new IllegalArgumentException("Sim device '" + aName + "' does not exist");
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Sim device '" << aName << "' does not exist");
    }

    return output;
}

class PigeonGyroWrapper : public AModuleWrapper, public IGyroWrapper
{
public:
    double mAngle{ 0 };
    double mAngleOffset{ 0 };

    static const std::string TYPE;
    std::string GetType() override
    {
        return TYPE;
    }

    explicit PigeonGyroWrapper(const std::string& aName) :
            AModuleWrapper(aName)
    {
    }

    // public void setDesiredYaw(double aOffset)
    // {
    //     mAngleOffset = aOffset - mAngle;
    // }

    double GetAngle() override
    {
        return mAngle + mAngleOffset;
    }

    void SetAngle(double aAngle) override
    {
        mAngle = aAngle;
    }
};

class PigeonAccelWrapper : public AModuleWrapper, public IAccelerometerWrapper
{
public:
    double mAccel{ 0 };

    explicit PigeonAccelWrapper(const std::string& aName) :
            AModuleWrapper(aName)
    {
    }

    double GetAcceleration() override
    {
        return mAccel;
    }

    void SetAcceleration(double aAccel) override
    {
        mAccel = aAccel;
    }
    static const std::string TYPE;
    std::string GetType() override
    {
        return TYPE;
    }
};

const std::string PigeonGyroWrapper::TYPE = "PigeonGyro";
const std::string PigeonAccelWrapper::TYPE = "PigeonAccel";

} // namespace

const std::string CtrePigeonImuSim::TYPE = "Pigeon";

CtrePigeonImuSim::CtrePigeonImuSim(int canHandle, int aBasePort) :
        AModuleWrapper("Pigeon IMU"),

        mYaw(new PigeonGyroWrapper("Pigeon Yaw")),
        mPitch(new PigeonGyroWrapper("Pigeon Pitch")),
        mRoll(new PigeonGyroWrapper("Pigeon Roll")),
        mX(new PigeonAccelWrapper("Pigeon X")),
        mY(new PigeonAccelWrapper("Pigeon Y")),
        mZ(new PigeonAccelWrapper("Pigeon Z")),
        mCanHandle(canHandle)

{
    SensorActuatorRegistry::Get().Register(aBasePort + 0, mX);
    SensorActuatorRegistry::Get().Register(aBasePort + 1, mY);
    SensorActuatorRegistry::Get().Register(aBasePort + 2, mZ);

    SensorActuatorRegistry::Get().Register(aBasePort + 0, mYaw);
    SensorActuatorRegistry::Get().Register(aBasePort + 1, mPitch);
    SensorActuatorRegistry::Get().Register(aBasePort + 2, mRoll);
}

void CtrePigeonImuSim::SetInitialized(bool aIsInitialized)
{
    AModuleWrapper::SetInitialized(aIsInitialized);

    mX->SetInitialized(aIsInitialized);
    mY->SetInitialized(aIsInitialized);
    mZ->SetInitialized(aIsInitialized);

    mYaw->SetInitialized(aIsInitialized);
    mPitch->SetInitialized(aIsInitialized);
    mRoll->SetInitialized(aIsInitialized);

    std::string deviceNum = std::to_string(mCanHandle);
    std::string deviceName = "CtrePigeonIMUWrapper " + deviceNum + "[" + deviceNum + "]";
    std::cout << "XXXX " << deviceName << std::endl;
    HAL_SimDeviceHandle deviceSimHandle{ 0 };
    frc::sim::SimDeviceSim::EnumerateDevices(
            deviceName.c_str(), [&](const char* name, HAL_SimDeviceHandle handle) {
                if (wpi::StringRef(name) == deviceName)
                {
                    deviceSimHandle = handle;
                }
            });

    if (deviceSimHandle)
    {
        mYawAngleDeg = getSimDouble(deviceSimHandle, "Yaw_angleDeg");
        mRawGyroXyzDps0 = getSimDouble(deviceSimHandle, "RawGyro_xyz_dps_0");
        mRawGyroXyzDps1 = getSimDouble(deviceSimHandle, "RawGyro_xyz_dps_1");
        mRawGyroXyzDps2 = getSimDouble(deviceSimHandle, "RawGyro_xyz_dps_2");
        mYawPitchRollYpr0 = getSimDouble(deviceSimHandle, "YawPitchRoll_ypr_0");
        mYawPitchRollYpr1 = getSimDouble(deviceSimHandle, "YawPitchRoll_ypr_1");
        mYawPitchRollYpr2 = getSimDouble(deviceSimHandle, "YawPitchRoll_ypr_2");
        mFusedHeading2Value = getSimDouble(deviceSimHandle, "FusedHeading1_value");
    }
}

std::shared_ptr<IGyroWrapper> CtrePigeonImuSim::getYawWrapper()
{
    return mYaw;
}
std::shared_ptr<IGyroWrapper> CtrePigeonImuSim::getPitchWrapper()
{
    return mPitch;
}
std::shared_ptr<IGyroWrapper> CtrePigeonImuSim::getRollWrapper()
{
    return mRoll;
}

std::shared_ptr<IAccelerometerWrapper> CtrePigeonImuSim::getXWrapper()
{
    return mX;
}
std::shared_ptr<IAccelerometerWrapper> CtrePigeonImuSim::getYWrapper()
{
    return mY;
}
std::shared_ptr<IAccelerometerWrapper> CtrePigeonImuSim::getZWrapper()
{
    return mZ;
}

///////////////////////////////////////////////
void CtrePigeonImuSim::handleSetYaw()
{
    SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Unsupported");
//    ((CtrePigeonImuSim.PigeonGyroWrapper) getYawWrapper()).setDesiredYaw(mYawAngleDeg.Get());
}


void CtrePigeonImuSim::handleGetRawGyro()
{
    mRawGyroXyzDps0.Set(getYawWrapper()->GetAngle());
    mRawGyroXyzDps1.Set(getPitchWrapper()->GetAngle());
    mRawGyroXyzDps2.Set(getRollWrapper()->GetAngle());
}

void CtrePigeonImuSim::handleGetYawPitchRoll()
{
    mYawPitchRollYpr0.Set(getYawWrapper()->GetAngle());
    mYawPitchRollYpr1.Set(getPitchWrapper()->GetAngle());
    mYawPitchRollYpr2.Set(getRollWrapper()->GetAngle());
}

void CtrePigeonImuSim::handleGetFusedHeading()
{
    mFusedHeading2Value.Set(getYawWrapper()->GetAngle());
}
///////////////////////////////////////////////
